### DISCUSSION TITLE
What is `waitForPendingLocaleChange()` supposed to do?

### DISCUSSION THREAD

**Discussion by frederikheld** [nuxt-modules/i18n#3161] [Q&A] [answered]
What is `waitForPendingLocaleChange()` supposed to do?

According to the docs it's this: "Returns a promise that will be resolved once the pending locale is set."

But it doesn't.

I want to change the route hash _after_* the locale was changed but if I use `await waitForPendingLocaleChange()` in front of `window.history.replaceState()` which I use to replace the hash, the url will jump back and forth between the routes and then get stuck in between.

If I wrap `window.history.replaceState()` in a `setTimeout()` of 100ms instead, it works.

So apparently `waitForPendingLocaleChange()` does not wait for anything :-|

\* preferably I'd like to change the hash in the same operation that changes the route prefix when changing the locale, but I did not find a way to hook into that operation.

**Comment by BobbieGoede** [comment, marked as answer]
The `waitForPendingLocaleChange()` function only does something if `skipSettingLocaleOnNavigate` is enabled, it's used when using page transitions, see https://i18n.nuxtjs.org/docs/guide/lang-switcher#wait-for-page-transition

**Reply by frederikheld** [reply]
Ah, so this is not what I'm looking for. Thanks for the explanation!

But how would I hook into the locale switch to have the locale prefix AND the route hash change at the same time?

**Reply by BobbieGoede** [reply]
I don't quite get the use case, but I feel like you should be able to achieve this using a middleware (https://nuxt.com/docs/guide/directory-structure/middleware), adding/redirecting to the path with hash when you detect a (locale) route change happening.

**Reply by frederikheld** [reply]
The use case is: my url hashes are localized, so I have to change the hash when the language is changed.

I tried it with middleware as well, but it does not work. It's difficult to filter out which route transition was caused by a language switch and which was caused by normal routing. Assuming, this can be solved without fault, the much bigger issue is that the replacement of the route hash triggers another route transition which causes a loop of half-finished route transitions. I tried it with `router.replace()` as well as with `window.history.replaceState()` and while the latter works a bit better, none of them work correctly. I never had such issues with "normal" routes, this just happens when I use prefixed routes with i18n, so I suppose that there's some magic going on with the i18n plugin.

I think this would be way easier if the i18n plugin would provide a hook to plug in middleware that can manipulate the transition. But I can't find any such hook.

**Reply by BobbieGoede** [reply]
Could you provide a minimal reproduction of you use case? I can take a look if I can get it working with a middleware. Surely it's possible to compare the locale of the previous and current route to determine if a language change is happening? 🤔

**Reply by frederikheld** [reply]
I'll set one up, but it will take some days. I've already settled with the less optimal (but working) approach, so I have to write the code again. Thanks for your help!
